Remove extraneous residual call at end of FEProblemBase::computeJacobian

lindsayad opened the pull request:
This call only touches MOOSE internals and is not useful for the Petsc
solve. All re-golded moose tests are simply counting residual evaluations in some
way. Contact related tests are expected to fail because the algorithm is not currently
consistent. We'll have to figure out a solution for this

Closes #10446

lindsayad committed:
Allow different residual functions for different residual contexts

- Remove -snes_mf and -snes_mffd from PetscSupport options because they will set the mffd function
  to our nonlinear residual computing function within the Petsc internals
- Change from using function pointers to functors for residuals per Roy Stogner suggestion
- When we receive a libmesh call-back for a non-linear residual evaluation we will set
  _computing_nonlinear_residual = true; then we can perform contact updates only for those residuals

lindsayad committed:
Remove extraneous residual call at end of FEProblemBase::computeJacobian

This call only touches MOOSE internals and is not useful for the Petsc
solve. All re-golded moose tests are simply counting residual evaluations in some
way. Contact related tests are expected to fail because the algorithm is not
consistent.

Closes #10446

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/148907/) on 63b5d95 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10597/site)

This comment will be updated on new commits.

lindsayad commented:
Failing tests:
```
combined/test:contact.4ElemTensionRelease_no_new_nonzeros ................................... FAILED (EXODIFF)
combined/test:contact_verification/patch_tests/automatic_patch_update.sliding_blocks ........ FAILED (EXODIFF)
combined/test:normalized_penalty.kin_Q4 ..................................................... FAILED (EXODIFF)
combined/test:normalized_penalty.kin_Q8 ..................................................... FAILED (EXODIFF)
combined/test:catch_release.test ............................................................ FAILED (EXODIFF)
combined/test:contact.4ElemTensionRelease_dirac ................................ [min_cpus=4] FAILED (EXODIFF)
combined/test:contact.4ElemTensionRelease_constraint ........................... [min_cpus=4] FAILED (EXODIFF)
combined/test:contact_verification/patch_tests/single_pnt_2d.mu_0_0_pen_sm ................... FAILED (ERRMSG)
combined/test:contact_verification/patch_tests/single_pnt_2d.glued_pen_sm .................... FAILED (ERRMSG)
combined/test:contact_verification/patch_tests/single_pnt_2d.frictionless_pen_sm ............. FAILED (ERRMSG)
combined/test:contact_verification/patch_tests/single_pnt_2d.mu_0_0_kin_sm ................... FAILED (ERRMSG)
combined/test:contact_verification/patch_tests/single_pnt_2d.mu_0_2_pen_sm ................... FAILED (ERRMSG)
combined/test:contact_verification/patch_tests/single_pnt_2d.frictionless_kin_sm ............. FAILED (ERRMSG)
combined/test:nodal_area.Hex20_3 ............................................................. FAILED (ERRMSG)
combined/test:nodal_area.Hex20 ............................................................... FAILED (ERRMSG)
combined/test:nodal_area.Hex27 ............................................................... FAILED (ERRMSG)
combined/test:generalized_plane_strain_tm_contact.generalized_plane_strain_tm_contact [min...] FAILED (ERRMSG)
```

lindsayad commented:
I don't know whether this was purposeful or not, but I believe this extra residual call actually saves mechanical contact (at least for the current algorithmic design). This design actually does allow the last non-linear residual to differ from the zeroth linear residual. It updates the residual to be in the same contact state as the Jacobian so that the linear solve makes sense. An example solve excerpt from current MOOSE:
```
 0 Nonlinear |R| = 1.156441e+04
    0 KSP unpreconditioned resid norm 1.156441422995e+04 true resid norm 1.156441422995e+04 ||r(i)||/||b|| 1.000000000000e+00
    1 KSP unpreconditioned resid norm 4.127321108483e+01 true resid norm 4.127321108483e+01 ||r(i)||/||b|| 3.568984149491e-03
    2 KSP unpreconditioned resid norm 5.942924286114e-01 true resid norm 5.942678624979e-01 ||r(i)||/||b|| 5.138763197871e-05
    3 KSP unpreconditioned resid norm 8.196865808799e-03 true resid norm 8.143681645983e-03 ||r(i)||/||b|| 7.042018284760e-07
 1 Nonlinear |R| = 4.159728e+01
    0 KSP unpreconditioned resid norm 2.489921379810e+05 true resid norm 2.489921379810e+05 ||r(i)||/||b|| 1.000000000000e+00
    1 KSP unpreconditioned resid norm 1.682255939819e+03 true resid norm 1.682255440180e+03 ||r(i)||/||b|| 6.756259269150e-03
    2 KSP unpreconditioned resid norm 3.475328940525e+02 true resid norm 3.475329851678e+02 ||r(i)||/||b|| 1.395758870083e-03
    3 KSP unpreconditioned resid norm 2.802467958181e+02 true resid norm 2.802470187309e+02 ||r(i)||/||b|| 1.125525572829e-03
    4 KSP unpreconditioned resid norm 2.618527049893e+02 true resid norm 2.618536041161e+02 ||r(i)||/||b|| 1.051654105384e-03
    5 KSP unpreconditioned resid norm 2.261539117820e+02 true resid norm 2.261552807474e+02 ||r(i)||/||b|| 9.082828180088e-04
    6 KSP unpreconditioned resid norm 1.861661249814e+02 true resid norm 1.861667621989e+02 ||r(i)||/||b|| 7.476812870818e-04
    7 KSP unpreconditioned resid norm 3.471621561051e+01 true resid norm 3.471608650988e+01 ||r(i)||/||b|| 1.394264364786e-04
    8 KSP unpreconditioned resid norm 2.787056786458e-02 true resid norm 2.843882954681e-02 ||r(i)||/||b|| 1.142157731461e-07
 2 Nonlinear |R| = 2.573203e+03
    0 KSP unpreconditioned resid norm 2.573203431198e+03 true resid norm 2.573203431198e+03 ||r(i)||/||b|| 1.000000000000e+00
    1 KSP unpreconditioned resid norm 1.696535044472e+02 true resid norm 1.696534382603e+02 ||r(i)||/||b|| 6.593083010981e-02
    2 KSP unpreconditioned resid norm 2.175582749333e+00 true resid norm 2.175437199198e+00 ||r(i)||/||b|| 8.454198268283e-04
    3 KSP unpreconditioned resid norm 5.426933937643e-03 true resid norm 5.442604806649e-03 ||r(i)||/||b|| 2.115108638773e-06
    4 KSP unpreconditioned resid norm 3.203630197162e-05 true resid norm 2.932633631562e-04 ||r(i)||/||b|| 1.139681999490e-07
 3 Nonlinear |R| = 5.835372e+02
```
@friedmud may be interested in this

lindsayad commented:
This saves contact at the expense of penalizing "regular" physics with an additional residual call for every non-linear iteration. I'm going to push to this PR with updates accompanying [this libmesh PR](https://github.com/libMesh/libmesh/pull/1553) that should allow the same consistency of linear solve for contact without penalizing other physics

lindsayad commented:
One could point out that the point of this call is explained fairly well in its comment 😝 .

However, I still think the new ability to discern between non-linear and other residual evaluations should make the current formulation unnecessary.

joshuahansel commented:
@andrsd I started looking into the RELAP-7 failures but haven't come up with much yet. I've looked through the constituent classes of `Turbine` but didn't find anything that I'd expect to change on a repeated residual evaluation. I'm doing some run comparisons to see if I can narrow down the problematic code.

lindsayad commented:
It also should be pointed out that we essentially catch the updated contact state one non-linear iteration late in the current formulation. This has consequences such as not capturing newly penetrated nodes on the last non-linear iteration as shown below:
![old-moose-penetration](https://user-images.githubusercontent.com/10248304/35395508-c77fe5cc-01a7-11e8-94dc-9b1d0776bfd5.png)

andrsd commented:
@joshuahansel Turbine is finicky, we will just need to play with solver settings. The other test might need the same.  So, the test failures in RELAP-7 are fine and we can deal with them after this is merged.

lindsayad committed:
Don't care about increasing nl residuals.

lindsayad committed:
Update contact set on nonlinear residuals in MechanicalContactState instead of Jacobians

- This is critical for several reasons:
    - If running JFNK or if running with a finite differenced preconditioner, the setupJacobian
      method will never get called
    - If updating on Jacobian, we may detect false convergence on the final non-linear iteration
      because if the Newton step would actually change the contact state, the nonlinear residual
      evaulation would have no idea. This is seen in a Ben Spencer example where new nodes come into
      contact on the final non-linear iteration but aren't caught; subsequently they penetrate the
      face

Other changes:

- Add a hook on residualEnd that prints out whether the contact state has changed since the last
  non-linear residual evaluation, and prints the number of nodes in contact
- Refactor MechanicalContactConstraint's updateContactSet, shouldApply, and computeContactForce
  methods so that we use an up-to-date contact force when determining release

lindsayad commented:
@fdkong Do you want to put a patch on top of this for copying in the residual vector for the first mffd iteration?

lindsayad commented:
Same contact problem with the new formulation:
![new-moose-penetration](https://user-images.githubusercontent.com/10248304/35416599-e9e46b30-01e6-11e8-982f-e2a6aa4756dd.png)

Note that slave penetration no long occurs.

fdkong commented on modules/contact/src/ContactMaster.C:
```diff
@@ -224,17 +180,22 @@ ContactMaster::addPoints()
       continue;
 
+    bool is_nonlinear =
+        _subproblem.getMooseApp().executioner()->feProblem().computingNonlinearResid();
```
Does not this `_subproblem.computingNonlinearResid()` work?

lindsayad commented:
No because for contact, subproblem is most likely referring to the displaced problem as opposed to the FEProblem

> On Jan 26, 2018, at 9:35 AM, Fande Kong <notifications@github.com> wrote:
> 
> @fdkong commented on this pull request.
> 
> In modules/contact/src/ContactMaster.C:
> 
> > @@ -224,17 +180,22 @@ ContactMaster::addPoints()
>      if (!pinfo || pinfo->_node->n_comp(_sys.number(), _vars[_component]) < 1)
>        continue;
>  
> +    bool is_nonlinear =
> +        _subproblem.getMooseApp().executioner()->feProblem().computingNonlinearResid();
> Does not this _subproblem.computingNonlinearResid() work?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

fdkong commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -149,7 +183,25 @@ NonlinearSystem::solve()
+    else
+      solver.solve_type() = PetscNonlinearSolver<Real>::MF;
+
+    solver.mffd_residual_object = &_fd_residual_functor;
```
Is it better to create a `_mffd_residual_functor`? For FDColoring, the residual evaluation could be called several times for a given linear iteration while it is called once for MFFD.

lindsayad replied:
This is moot now with the form of the Petsc API changes

lindsayad commented:
The coloring residual calls happen before KSPSolve right? And we get a number of residual calls equal to the number of colors? Plus for coloring I'm guessing we would run into the same issue that we're seeing with mffd where there would be an extra residual call at the beginning of the coloring process to compute the un-perturbed residual.

So if we had different objects (say _mffd_residual_functor and _fd_residual_functor), could we apply different checks for each where we could determine whether we're computing the base un-perturbed residual and then copy in the previously computed non-linear residual?

> On Jan 26, 2018, at 10:52 AM, Fande Kong <notifications@github.com> wrote:
> 
> @fdkong commented on this pull request.
> 
> In framework/src/base/NonlinearSystem.C:
> 
> >      setupFiniteDifferencedPreconditioner();
> +  }
> +
> +#ifdef LIBMESH_HAVE_PETSC
> +  PetscNonlinearSolver<Real> & solver =
> +      static_cast<PetscNonlinearSolver<Real> &>(*_transient_sys.nonlinear_solver);
> +  Moose::SolveType moose_solve_type = _fe_problem.solverParams()._type;
> +  if (moose_solve_type == Moose::ST_PJFNK || moose_solve_type == Moose::ST_JFNK)
> +  {
> +    if (moose_solve_type == Moose::ST_PJFNK)
> +      solver.solve_type() = PetscNonlinearSolver<Real>::MF_OPERATOR;
> +    else
> +      solver.solve_type() = PetscNonlinearSolver<Real>::MF;
> +
> +    solver.mffd_residual_object = &_fd_residual_functor;
> Is it better to create a _mffd_residual_functor? For FDColoring, the residual evaluation could be called several times for a given linear iteration while it is called once for MFFD.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

lindsayad commented:
Note that this PR alone does not reduce the total number of residual evaluations because now that we're using different functions for `SNES` and `MFFD` current Petsc thinks that it has to compute an independent base residual for its finite differencing. However, after this [Petsc PR](https://bitbucket.org/petsc/petsc/pull-requests/845/added-matsnesmfsetreusebase-at-request-of/diff) and this [Libmesh PR](https://github.com/libMesh/libmesh/pull/1572#discussion_r164561268), we'll have successfully eliminated one residual evaluation per non-linear iteration.

fdkong commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -149,7 +144,25 @@ NonlinearSystem::solve()
+  PetscNonlinearSolver<Real> & solver =
+      static_cast<PetscNonlinearSolver<Real> &>(*_transient_sys.nonlinear_solver);
+  Moose::SolveType moose_solve_type = _fe_problem.solverParams()._type;
+  if (moose_solve_type == Moose::ST_PJFNK || moose_solve_type == Moose::ST_JFNK)
```
There is a situation, where code may not work properly.  If we setup a matrix free method through a PETSc command-line option "`-snes_mf_operator 1`", and then the things in MOOSE and Libmesh may not be setup properly.

lindsayad replied:
If `-snes_mf_operator 1` is given in the input file, we already warn users that they should not do that (see [here](https://github.com/idaholab/moose/blob/devel/framework/src/utils/PetscSupport.C#L655). If it is specified on the command line, then that would only affect contact.

lindsayad replied:
I had thought of this before, but I don't have a good suggestion for fixing it...

lindsayad replied:
So @fdkong, per our discussion yesterday, this is something that should be fixable by reordering/refactor the `PetscNonlinearSolver::solve` in libmesh

fdkong replied:
Pushed changes up to Libmesh at https://github.com/libMesh/libmesh/pull/1572/files

lindsayad commented:
@YaqiWang This fails a Rattlesnake test although the Exodiff doesn't appear to be too significant to me. What do you want to do with it?

YaqiWang commented:
You can try set a tighter tolerance without this PR and regold to that test to see if Rattlesnake with pass with this PR. I suppose this will fix it.

fdkong commented on modules/combined/test/tests/contact_verification/patch_tests/plane_3/plane3_template1.i:
```diff
@@ -263,8 +263,8 @@
-  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
-  petsc_options_value = 'lu       superlu_dist'
+  petsc_options_iname = '-pc_type'
+  petsc_options_value = 'asm'
```
Does LU throw out errors on this?

lindsayad replied:
It did on linux but ran fine on Mac. I would like to re-run this on rod without superlu_dist

lindsayad replied:
Yea, so I just ran this on cone, and with `superlu_dist` we get this solve behavior:
```
 0 Nonlinear |R| = 7.770396e+01
Solution, reference convergence variable norms:
disp_x: 0  saved_x: 0
disp_y: 77.704  saved_y: 0
0       Linear |R| = 7.770396e+01
1       Linear |R| = 7.770171e+01
2       Linear |R| = 1.062272e+01
3       Linear |R| = 2.589127e+00
4       Linear |R| = 1.132553e+00
5       Linear |R| = 2.354931e-01
6       Linear |R| = 1.057685e-01
7       Linear |R| = 4.257440e-02
1  Nonlinear |R| = 3.262283e+03
```
and the solve eventually fails with the message: `Failure in NearestNodeThread because solution contains inf or not-a-number entries.  This is likely due to a failed factorization of the Jacobian matrix.`

If I no longer use superlu_dist (instead using petsc's lu), the the solve succeeds easily:
```
 0 Nonlinear |R| = 7.770396e+01
Solution, reference convergence variable norms:
disp_x: 0  saved_x: 0
disp_y: 77.704  saved_y: 0
0       Linear |R| = 7.770396e+01
1       Linear |R| = 2.253578e+01
2       Linear |R| = 2.626581e-11
1  Nonlinear |R| = 3.330550e-03
```
So it looks like there's still some buggy behavior in `superlu_dist` as @bwspenc has experienced before

lindsayad commented:
@aeslaughter Are you familiar with the failing mastodon test?

aeslaughter commented:
@lindsayad I am not familiar with the test, but it is one of our ludicrous tests that have 100s of timesteps, I am guessing that there is a slight change in one of the timesteps and it throws something off.

@sveerara @cbolisetti  Do you think we can update these tests to output at specific sync times? If so, what times would work. We can re-gold with sync times using current devel and see if that will fix this problem. Let me know and I can do this today.

sveerara commented:
@aeslaughter, it does look like dt gets halved at some time step and the exodus files are different from that point in time. If the output sync times can be made same as dt provided in the input file (0.01 s), that would probably fix this problem. Thanks!

lindsayad commented:
@bwspenc These three regolded files:
- `incremental_slip_out.e`
- `glued_contact/*/mechanical_constraint_out.e`
- `glued_contact/*/out.e`

all exhibited delayed contact, e.g. contact is now detected correctly on the last non-linear iteration whereas before contact wasn't detected until the beginning of the next time step.

- `8ElemTensionRelease` and `sliding_update_out` looks about the same to me (looking at displacements, penetration, and contact pressure)

moosebuild commented:
Job [App tests](https://www.moosebuild.org/job/150823/) on 7b54908 : invalidated by @lindsayad

Merged patches to Bison, Mastodon, and Marmot

brianmoose commented:
Looks like this caused the 64 bit dof build to fail: https://www.moosebuild.org/job/151374/
refs #10655

brianmoose commented:
Hmm, looks like this also introduced a memory leak: https://www.moosebuild.org/job/151369/
refs #10657

lindsayad commented:
Ugh, this was introduced by libmesh/libmesh#1553. Needed a `MatDestroy(&J);`

brianmoose commented:
I am guessing this one caused the heavy test failures as well: https://www.moosebuild.org/job/151938/
issue #10665
